### DISCUSSION TITLE
Best Practices for Realtime UI updates in Unity

### DISCUSSION THREAD

**Discussion by GrimLothar** [supabase-community/supabase-csharp#126] [Q&A] [answered]
I'm using the library in Unity and it's working pretty well! My question might be more Unity oriented, but since i'm specifically having this problem only with this library, I figured I would ask here. 

Basically I'm subscribing to some Realtime channels in order to update my game's UI when the DB changes. The problem i'm facing is that Unity can't update the main thread from a different thread which seems to be the case whenever I get updates from the socket.

The way I'm currently solving it is like this:
```c#
public class SupabaseManager : MonoBehaviour
{
    public UnityAction<string, string> onMatchRequest;

    private string SupabaseURL = "";
    private string SupabasePublicKey = "";

    private Client _supabase;

    private RealtimeBroadcast<UserBroadcast> _userBroadcast;
    private static SupabaseManager instance;
    private SynchronizationContext _mainThreadContext;

    private readonly Dictionary<string, RealtimeChannel> channels = new Dictionary<string, RealtimeChannel>();

    private void Awake()
    {
        instance = this;
        SupabaseURL = NetworkData.GetSupabaseUrl();
        SupabasePublicKey = NetworkData.GetSupabasePublicKey();
    }
    
    async void Start()
    {
        _mainThreadContext = SynchronizationContext.Current;
        string url = SupabaseURL;
        string key = SupabasePublicKey;

        if (_supabase == null)
        {
            var options = new SupabaseOptions
            {
                Headers = new Dictionary<string, string>
                {
                    ["Authorization"] = "Bearer " + ApiClient.Get().AccessToken
                }
            };
            _supabase = new Client(url, key, options);
            await _supabase.InitializeAsync();
        }

        _supabase.Realtime.AddDebugHandler((sender, message, exception) => Debug.Log(message));
        _supabase.Realtime.SetAuth(ApiClient.Get().AccessToken);
        var realtimeClient = await _supabase.Realtime.ConnectAsync();
    }

    private void OnDestroy()
    {
        Debug.Log("SupabaseManager OnDestroy");
        if (_supabase != null)
        {
            foreach (var channel in channels.Values)
            {
                _supabase.Realtime.Remove(channel);
            }
            channels.Clear();
            
            _supabase.Realtime.RemoveStateChangedHandler(SocketEventHandler);
            _supabase.Realtime.Disconnect();
            _supabase = null;
        }
    }

    public delegate void TournamentChangedHandler(Tournament tournament);
    public async void SubscribeToTournamentChanges(int tournament_id, TournamentChangedHandler handler)
    {
        var channelName = "tournaments:" + tournament_id;
        channels.TryGetValue(channelName, out RealtimeChannel channel);
        if (channel == null)
        {
            channel = _supabase.Realtime.Channel(channelName);
            channels.Add(channelName, channel);
            channel.Register(new PostgresChangesOptions("public", "aw-tournaments", ListenType.Updates, "id=eq." + tournament_id ));
            channel.AddPostgresChangeHandler(ListenType.Updates, (_, change) =>
            {
                if (change.Payload?.Data?.Table == "aw-tournaments")
                {
                    //Tournament data changed
                    var model = change.Model<Tournament>();
                    if (model != null)
                    {
                        _mainThreadContext.Post(_ => handler.Invoke(model), null);
                    }
                }
            });
        }
        await channel.Subscribe();
    }

    public void UnsubscribeToTournamentChanges(int tournament_id)
    {
        var channelName = "tournaments:" + tournament_id;
        channels.TryGetValue(channelName, out RealtimeChannel channel);
        if (channel != null)
        {
            channel.Unsubscribe();
        }
    }

    public static SupabaseManager Get()
    {
        return instance;
    }

}

```

I basically have a `SynchronizationContext _mainThreadContext` which I set to `Current` on `Start` (Start gets called on the Main Thread)` and whenever I get an update, i then call whatever method I need using this `_mainThreadContext` like this:  `_mainThreadContext.Post(_ => handler.Invoke(model), null);`

It seems to be working, but not sure if this is the best way to do this?

Would love to hear other options or validation that this is correct, since I don't have a lot of experience with multi threading.

Thanks!

**Comment by wiverson** [comment, marked as answer]
Yeah, it's a PITA.

Here's the basic pattern for what I wind up doing. Let's say the thing I'm interested in is messages from the server. So I'd have a var incoming = new List<Message> in my MonoBehavior. Then I add messages to this List via the channel updates. Then I have an Update() method that looks for messages and removes them, which is fine as Update always runs on the UI thread. I use UniTask to make sure the async/await is managed properly WRT to setting up the context.

I like your version as well, it's probably more technically accurate but I think it's basically doing the same thing WRT setting up the sync context as UniTask & Update.

**Reply by acupofjose** [reply]
Is there a way to restructure to make this _not_ a PITA?

**Reply by wiverson** [reply]
I think a Unity specific helper class might be useful, maybe? It would just be thin data wrapper for either of the two patterns (grimAgent's or mine), but the helper would be very Unity specific.

Better would be for Unity to fix their cruddy async/await implementation. Unity has been investing in a ton of stuff that's not done and neglecting their core for a long, long time now. For example, investing a ton of energy in DOTS but not finishing async/await for core MonoBehavior. Lagging way behind their .NET implementation, a wildly hacked up version with a ton of weird stuff. Having a custom fork of NUnit with only partial support for async await (e.g. you can make a individual test case async but before/after async will crash).

FWIW Godot has a much cleaner .NET impl...
